File: baseunits/modules/YoungAceUp.pas
unit YoungAceUp;

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, RegExpr, Dialogs;

implementation

const
  dirurl = '/comics/';

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + dirurl) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        for v in XPath('//ul[@class="manga-list__list"]/li/h4/a') do
        begin
          ALinks.Add(v.toNode.getAttribute('href'));
          ANames.Add(v.toString);
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  s, n: String;
  v: IXQValue;
  i: Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := AppendURLDelim(MaybeFillHost(Module.RootURL, AURL));
    if GET(url) then
    begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := MaybeFillHost(Module.RootURL, XPathString('//*[@class="fancybox a-alpha"]/img/@src'));
          if title = '' then title := XPathString('//strong');
          summary := XPathString('//*[@class="single-story"]/p');
          { there is no chapter list?
            assuming the first chapter link in manga info is always the last chapters }
            for v in XPath('//a[@class="single"]') do
            begin
              chapterLinks.Add(v.toNode.getAttribute('href'));
              chapterName.Add(v.toString);
            end
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  v: IXQValue;
  ViewerURL: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    ViewerURL := 'http://viewer.tonarinoyj.jp';
    if GET(FillHost(ViewerURL, AURL)) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          for v in XPath('//img[@class="js-page-image"]') do
            PageLinks.Add(v.toNode.getAttribute('src'));
        finally
          Free;
        end;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'YoungAceUp';
    RootURL := 'https://web-ace.jp/youngaceup';
    Category := 'Raw';
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/YoungAceUp.lua
function GetNameAndLink()
	if http.get(module.rooturl..'/comics/') then
	    TXQuery.Create(http.Document).XPathHREFAll('//ul[@class="manga-list__list"]/li/h4/a', links, names)
		return no_error
	else
		return net_problem
	end
end

function GetInfo()
	mangainfo.url=AppendURLDelim(MaybeFillHost(module.rooturl,url))
	if http.get(mangainfo.url) then
		x=TXQuery.Create(http.Document)
		mangainfo.title=x.XPathString('//strong')
		mangainfo.coverlink=x.XPathString('//*[@class="fancybox a-alpha"]/img/@src')
		mangainfo.summary=x.XPathString('//*[@class="single-story"]/p')
		-- there is no chapter list?
		-- assuming the first chapter link in manga info is always the last chapters
	    x.XPathHREFAll('//a[@class="single"]', mangainfo.chapterlinks, mangainfo.chapternames)
		return no_error
	else
		return net_problem
	end
end

function GetPageNumber()
	if http.get(FillHost('http://viewer.tonarinoyj.jp',url)) then
		TXQuery.Create(http.document).XPathStringAll('//img[@class="js-page-image"]/@src', task.PageLinks)
		return true
	else
		return false
	end
end

function Init()
	m=NewModule()
	m.Category='Raw'
	m.Website='YoungAceUp'
	m.RootURL='https://web-ace.jp/youngaceup'
	m.OnGetNameAndLink='GetNameAndLink'
	m.OnGetInfo='GetInfo'
	m.OnGetPageNumber='GetPageNumber'
end
